In [28]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [29]:
data = pd.read_csv('../data/crop_recommendation.csv')
data

,N,P,K,temperature,humidity,ph,rainfall,label
0,90,42,43,20.879744,82.002744,6.502985,202.935536,rice
1,85,58,41,21.770462,80.319644,7.038096,226.655537,rice
2,60,55,44,23.004459,82.320763,7.840207,263.964248,rice
3,74,35,40,26.491096,80.158363,6.980401,242.864034,rice
4,78,42,42,20.130175,81.604873,7.628473,262.717340,rice
...,...,...,...,...,...,...,...,...
2195,107,34,32,26.774637,66.413269,6.780064,177.774507,coffee
2196,99,15,27,27.417112,56.636362,6.086922,127.924610,coffee
2197,118,33,30,24.131797,67.225123,6.362608,173.322839,coffee
2198,117,32,34,26.272418,52.127394,6.758793,127.175293,coffee


In [30]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier


In [31]:
# definir variables predictoras y variable objetivo

X = data.drop(columns=['label'])
y = data.label
X.shape, y.shape

((2200, 7), (2200,))

In [32]:
# definir pipeline y parámetros para GridSearchCV

pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('model', KNeighborsClassifier())
])

params = {
        "model__n_neighbors": [3, 5, 7, 9,11, 13, 15],
        "model__weights": ["uniform", "distance"],
        "model__metric": ["euclidean", "manhattan", "minkowski"]
}


In [33]:
# split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Grid Search para encontrar los mejores parámetros del modelo
grid_search = GridSearchCV(pipeline, params, cv=5, scoring="f1_macro", n_jobs=12, verbose = 1)
grid_search.fit(X_train, y_train)
display(grid_search.best_params_)
display(grid_search.best_score_)
pipeline = grid_search.best_estimator_

"""
Output:

"""

Fitting 5 folds for each of 42 candidates, totalling 210 fits


{'model__metric': 'manhattan',
 'model__n_neighbors': 3,
 'model__weights': 'distance'}

np.float64(0.9828923060172319)

'\nOutput:\n\n'

In [34]:
from sklearn.metrics import classification_report, accuracy_score
#predicciones
y_pred = pipeline.predict(X_test)

#mostrar resultados
print(classification_report(y_test, y_pred))

#accuracy 
print("Accuracy:", accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       1.00      0.95      0.97        20
    chickpea       1.00      1.00      1.00        26
     coconut       1.00      1.00      1.00        27
      coffee       1.00      1.00      1.00        17
      cotton       0.94      1.00      0.97        17
      grapes       1.00      1.00      1.00        14
        jute       0.88      1.00      0.94        23
 kidneybeans       0.91      1.00      0.95        20
      lentil       0.73      1.00      0.85        11
       maize       1.00      0.95      0.98        21
       mango       0.95      1.00      0.97        19
   mothbeans       1.00      0.88      0.93        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange       1.00      1.00      1.00        14
      papaya       1.00    

## Resultados



| Scoring en CV | Desempeño promedio en CV | Accuracy obtenida |
| ---------------- | ----- | -------- |
| Accuracy | 0.9829 | 0.975 |
| f1_macro | 0.9828 | 0.975 | 
| roc_auc_ovr | 0.9999 | 0.959 |

